pg-extend/src/lib.rs

```diff
@@ -64,40 +67,104 @@ macro_rules! pg_magic {
-    let args_null =
-        func_call_info.argnull[..num_args]
+    let args_null = func_call_info.argnull[..num_args]
         .iter()
         .map(|b| pg_bool::Bool::from(*b));
 
     (args, args_null)
 }
 
+struct PanicContext {
+    jump_value: c_int,
+    stack: *mut c_int,
+}
+
+unsafe impl Send for PanicContext {}
+
 /// This will replace the current panic_handler
 pub fn register_panic_handler() {
     use std::panic;
 
     // set (and replace the existing) panic handler, this will tell Postgres that the call failed
     //   a level of Fatal will force the DB connection to be killed.
     panic::set_hook(Box::new(|info| {
-        let level = pg_error::Level::Fatal;
+        // downcast info, check if it's the value we need.
+        //   this must check if the panic was due to a longjmp
+        //   the fence is to make sure the longjmp is not reodered.
+        compiler_fence(Ordering::SeqCst);
+        if let Some(panic_context) = info.payload().downcast_ref::<PanicContext>() {
+            // WARNING: do not set this level above Notice (ERROR, FATAL, PANIC), as it will calse
+            //   the following longjmp to execute.
+            let level = pg_error::Level::Notice;
+            pg_error::log(
+                level,
+                file!(),
+                line!(),
+                module_path!(),
+                format!("continuing longjmp: {}", info),
+            );
+            
+            // the panic came from a pg longjmp... so unwrap it and rethrow
+            unsafe {
+                pg_sys::longjmp(panic_context.stack, panic_context.jump_value);
+            }
+        } else {
+            // error level will cause a longjmp in Postgres
+            let level = pg_error::Level::Error;
+            pg_error::log(
+                level,
+                file!(),
+                line!(),
+                module_path!(),
+                format!("panic in Rust extension: {}", info),
+            );
+        }
 
-        pg_error::log(level, file!(), line!(), module_path!(), format!("panic in Rust extension: {}", info));
+        unreachable!("all above statements should have cause a longjmp to Postgres");
     }));
 }
 
-/// auto generate function to output a SQL create statement for the function
+/// Provides a barrier between Rust and Postgres' usage of the C set/longjmp
 /// 
+/// In the case of a longjmp being caught, this will convert that to a panic. For this to work
+///   properly, there must be a Rust panic handler (see crate::register_panic_handler).PanicContext
+///   If the `pg_exern` attribute macro is used for exposing Rust functions to Postgres, then
+///   this is already handled.
+///
+/// See the man pages for info on setjmp http://man7.org/linux/man-pages/man3/setjmp.3.html
+#[inline(never)]
+pub(crate) unsafe fn guard_c<R, F: FnOnce() -> R>(f: F) -> R {
+    // setup the check protection
+    let jumped = pg_sys::setjmp(pg_sys::PG_exception_stack as *mut std::os::raw::c_int);
```

Comment:
I think this might be wrong. This should be a temporary local pointer that's uninitialized, and then we can just longjmp with the same pointer to whatever stack frame was passed in.

Reply:
Ok, yes. Here's the plan, we need to preserve each of these setjmps:

sigint_interrupt_jmp
illegal_instruction_jump
GUC_flex_fatal_jmp

In addition to the one already caught. We need to set aside local space for the setjmp buffer that already exists in that location, copy the data from the global to the local. Then panic with the pointer to the proper global. One generic function should work for each setjmp check.